File: tourist-spots/park.js
const Stack = require('stackjs');
const Queue = require('jsQueue');
const readline = require('readline');
const conn = require('./MySQLConfig')();


class car {
	constructor(arrive_time,plate_number) {
		this.arrive_time = arrive_time;
		this.plate_number = plate_number;
		this.enter_time = null;
		this.leave_time = null;
	}

	copy() {
		var c = new car(this.arrive_time,this.plate_number);
		return c;
	}

	equals(car) {
		return (this.plate_number == car.plate_number);
	}

	leave() {
		this.leave_time = (new Date()).valueOf();
		console.log(this.plate_number,"leaved.");
		var cost = this.leave_time-this.enter_time;
		console.log("cost time: ",cost);
		conn.query(`insert into parkLog (plate_number,arrive_time,enter_time,leave_time,cost) values ("${this.plate_number}",${this.arrive_time},${this.enter_time},${this.leave_time},${cost})`,(err,data)=>{
			if(err) console.log('save log err:',err);
		});
		return cost;
	}
}

class park {
	constructor(park_capacity) {
		this.park_capacity = park_capacity;

		this.park_zone = new Stack();
		this.wait_zone = new Queue();
		this.temp_zone = new Stack();
	}

	park_car(plate_number) {
		var c = new car((new Date()).valueOf(),plate_number);
		var status = this.check_status(c);
		if(status<0) {
			if(status == -1) console.log("操作失败，车已经停在车道上了");
			else if(status == -2) console.log("操作失败，车已经停在车库中了");
			return false;
		}
		else {
			if(this.park_zone.size()>=this.park_capacity) {
				c.enter_time = (new Date()).valueOf();
				this.wait_zone.enqueue(c);
				console.log("操作成功,",c.plate_number,",车停入了等待区，位置 W:",this.wait_zone.size()-1);
			}
			else {
				c.enter_time = (new Date()).valueOf();
				this.park_zone.push(c);
				c.arrive_time = (new Date()).valueOf();
				console.log("操作成功,",c.plate_number,"车停入了停车场，位置 P:",this.park_zone.size()-1);
			}
			//this.display();//
			return true;
		}
	}


	stack_has_element(stack,element) {//whether the stack has the element(car)
		var temp_stack = new Stack();
		var ans = false;
		while(!stack.isEmpty()) {
			var t = stack.peek();
			if(t.equals(element)) {
				ans = true;
				break;
			}
			stack.pop();
			temp_stack.push(t);
		}
		while(!temp_stack.isEmpty()) {
			stack.push(temp_stack.peek());
			temp_stack.pop();
		}
		return ans;
	}


	queue_has_element(queue,element) {//whether the queue has the element(car)
		var temp_queue = new Queue();
		var ans = false;
		while(queue.size()!=0) {
			var t = queue.peek();
			if(t.equals(element)) {
				ans = true;
				break;
			}
			queue.dequeue();
			temp_queue.enqueue(t);
		}
		while(temp_queue.size()!=0) {
			queue.enqueue(temp_queue.peek());
			temp_queue.dequeue();
		}
		return ans;
	}


	check_status(c) {
		/*
		return -1:车已经停在车道上等待了
		return -2:车已经停在车库中了
		return  0:车已经离开或者还没到
		*/
		var ts = new Stack();
		if(this.queue_has_element(this.wait_zone,c)) {
			console.log("车已经停在车道上等待了");
			return -1;
		}
		else if(this.stack_has_element(this.park_zone,c)){
			console.log("车已经停在车库中了");
			return -2;
		}
		else return 0;
	}

	left_car(plate_number) {
		var time_now = (new Date()).valueOf();
		var c = new car(time_now,plate_number);
		var status = this.check_status(c);
		switch (status) {
			case 0://not in
				console.log("这辆车还没来");
				return {
					res:false,
					msg:'这辆车还没来'
				}
				break;
			case -1://in waiting zone
				console.log("不支持等待区离开");
				return {
					res:false,
					msg:'不支持等待区离开'
				}
				break;
			case -2://in park zon
				while (!this.park_zone.peek().equals(c)) {
					this.temp_zone.push(this.park_zone.peek());
					this.park_zone.pop();
				}

				var cost = this.park_zone.peek().leave();
				this.park_zone.pop();

				while (!this.temp_zone.isEmpty()) {
					this.park_zone.push(this.temp_zone.peek());
					this.temp_zone.pop();
				}

				if(this.wait_zone.size()!=0) {
					this.wait_zone.peek().enter_time = (new Date()).valueOf();
					this.park_zone.push(this.wait_zone.peek());
					this.wait_zone.dequeue();
				}
				//display();//
				return {
					res:true,
					cost:cost
				}
				break;
			default:
				console.log("car_status_err!");
				exit(1);
				break;
		}
	}

	display() {
		var temp_queue = new Queue();
		var temp_stack = new Stack();
		console.log("停车场状态:\n");
		console.log("等待区:");
		var k = 0;

		while(this.wait_zone.size()!=0) {
			//str += ("["+(k++)+"]"+this.wait_zone.peek()+" ");
			console.log(this.wait_zone.peek());
			temp_queue.enqueue(this.wait_zone.peek());
			this.wait_zone.dequeue();
		}
		while(temp_queue.size()!=0) {
			this.wait_zone.enqueue(temp_queue.peek());
			temp_queue.dequeue();
		}


		console.log("停车区:");
		while(!this.park_zone.isEmpty()) {
			temp_stack.push(this.park_zone.peek());
			this.park_zone.pop();
		}
		var k =0;
		while(!temp_stack.isEmpty()) {
			console.log(temp_stack.peek());
			//str += ("["+(k++)+"]"+temp_stack.peek()+" ");
			this.park_zone.push(temp_stack.peek());
			temp_stack.pop();
		}
		console.log("-------------------------------------------------------------");
	}

}


module.exports = function (park_size) { //init park to p and return
  var p = new park(park_size);
	console.log('park initialized...');
	/*
  p.park_car("wxx1");//p.display();
  p.park_car("wxx2");//p.display();
*/
  return p;
}

File: tourist-spots/cpp/park.cpp
#include <iostream>
#include <stack>
#include <ctime>
#include <queue>
#include <string>
using namespace std;

#define INF 99999
int park_capacity = 5;

struct car {
	time_t arrive_time;
	time_t enter_time;
	string plate_number;
	
	car(time_t _arrive_time,string _plate_num):
		arrive_time(_arrive_time),plate_number(_plate_num) {
			time(&enter_time);
	}
	
	car& operator = (car &c) {
		arrive_time = c.arrive_time;
		enter_time = c.enter_time;
		plate_number = c.plate_number;
		return c;
	}
	
	bool operator == (car& c) {
		return c.plate_number==plate_number;
	}
	
	bool operator != (car& c) {
		return c.plate_number!=plate_number;
	}
	
	void leave() {
		cout << plate_number << "离开了停车场，停车时间: "<< time(0)-enter_time<< "s" << endl;
	}
};


class park{
	public:
	
	stack<car> park_zone;
	queue<car> wait_zone;
	stack<car> temp_zone;
	
	public:
		void park_car(string plate_num,time_t _time) {
			car c(_time,plate_num);
			int status = check(c);
			if(status) {
				return;
				cout << (status==1?"车已经停在车库中了":"车已经停在车道上了") << endl;
			} 
			else {
				if(park_zone.size()>=park_capacity) {
					c.enter_time = time(0);
					wait_zone.push(c);
					cout << c.plate_number << "车停入了等待区，位置 W：";
					cout << wait_zone.size() << endl;
				}
				else {
					c.enter_time = time(0);
					park_zone.push(c);
					c.arrive_time = time(0);
					cout << c.plate_number << " 车停入了停车场，位置 P：";
					cout << park_zone.size() << endl;
				}
				display();//
			}
		}
		
		void park_car(string plate_num) {
			park_car(plate_num,time(0));
			
		}
		
		int check(car &c) { //return whether car has arrived
			stack<car> pz = park_zone;
			queue<car> wz = wait_zone;
			while (!pz.empty()) {
				if(pz.top()==c) {
					//cout << "车已经停在车库中了" << endl;
					return 1;
				}
				pz.pop();
			}
			while (!wz.empty()) {
				if(wz.front()==c) {
					//cout << "车已经停在车道上了" << endl;
					return 2;
				}
				wz.pop();
			}
			return 0;
		}
		
		void left_car(string plate_num) {
			left_car(plate_num,time(0));
		}
		
		
		void left_car(string plate_num,time_t _time) {
			car c(_time,plate_num);
			int status = check(c);
			switch (status) {
				case 0://not in
					cout << "这辆车还没来" << endl;
					break;
				case 2://in waiting zone
					cout << "不支持等待区离开" << endl;
					break;
				case 1://in park zone
					while (park_zone.top()!=c) {
						temp_zone.push(park_zone.top());
						park_zone.pop();
					}
					
					park_zone.top().leave();
					park_zone.pop();
					
					while (!temp_zone.empty()) {
						park_zone.push(temp_zone.top());
						temp_zone.pop();
					}
					
					if(wait_zone.size()>0) {
						wait_zone.front().enter_time = time(0);
						park_zone.push(wait_zone.front());
						wait_zone.pop();
					}
					display();//
					break;
			}
		}
		
		void display() {
			queue<car> wz = wait_zone;
			stack<car> pz = park_zone;
			
			cout << "停车场状态： " << endl ;
			int i=1;
			while (!pz.empty()) {
				cout << "["<< i++ << "]" << pz.top().plate_number << " ";
				pz.pop();
			}
			i=1;
			cout << "\n等待区状态： " << endl ;
			while (!wz.empty()) {
				cout << "["<< i++ << "]" << wz.front().plate_number << " ";
				wz.pop();
			}
			cout << endl;
		}

};

int main(int argc, char *argv[]) {
	park p;
	
	for(int i=0;i<5;i++) {
		string c;
		cin >> c;
		p.park_car(c);
	}
	for(int i=0;i<3;i++) {
		string c;
		cin >> c;
		p.left_car(c);
	}
	

	
	return 0;
}
